### ISSUE TITLE
Dynamic Collision Checking is Broken

### ISSUE THREAD

**Issue by surajRathi** [Team-Abhiyaan/iac_planner#5]
in branch [`collision_rework`](https://github.com/Team-Abhiyaan/iac_planner/tree/collision_rework),  looking at file [`test/collision_check.py`](https://github.com/Team-Abhiyaan/iac_planner/blob/collision_rework/test/collision_check.py)
The example has an other_vehicle at (1,0) in local frame with zero velocity, essentially parked in front of our ego vehicle. And this should block all paths.

However, despite non-zero ego velocity profile, the green paths are found to be valid.

![image](https://user-images.githubusercontent.com/6284428/114441268-cc76fc80-9be8-11eb-905e-1b4e07dfd584.png)

Here the ego vehicle is in blue, the lane boundaries are also in blue. The other vehicle is in black.

The paths found to be valid are in green, and the rejected paths (due to collision) are in red.

Is this a legitimate bug or is it only due to the other_vehicle having low speed?

See commit: 87247489cfdffb06d08397f3d6c3bdc8cc5cd656

**Comment by surajRathi** [comment]
Removing the  growth_factor  in 59c7c5806f4be7b6a7cd907d1ce4060b6ff122a6 has semi fixed the issue.

![image](https://user-images.githubusercontent.com/6284428/114500931-a41deb00-9c46-11eb-915c-f03cfb3385f7.png)

**Comment by surajRathi** [comment]
TODO: Implement first part of

> We perform this collision-checking efficiently using ahierarchical approach. Given a candidate trajectory for ourvehicle and a predicted trajectory for a dynamic obsta-cle (extended out in time to match the time duration ofthe candidate trajectory), we first construct a conservativebounding box for each trajectory. These bounding boxesrepresent a pessimistic approximation of the area of theenvironment the trajectories encounter. We then check to seeif these bounding boxes overlap: if they don’t, then the twotrajectories cannot intersect each other; if they do, then thereis a chance the trajectories intersect and we must continueto investigate. We then take the two trajectories and stepalong them in chronological, synchronized time. At eachtime instanttiwe construct a pessimistic bounding circle of the extent of our vehicle and the dynamic obstacle andcheck if these circles intersect. If we reach the end of ourtrajectory without any such intersections, the two trajectoriescannot intersect each other1. If the circles intersect at sometimetk, then we construct accurate polygonal representationsof our vehicle and the dynamic obstacle at this timetkand check if these polygons intersect. If so, the trajectorieswill collide with each other and this candidate trajectoryis removed from contention. If not, we continue to stepforwards in time performing our pessimistic bounding circlechecks. We continue in this fashion until we reach the endof the candidate trajectory.

from [here](https://www.ri.cmu.edu/pub_files/pub4/ferguson_david_2008_3/ferguson_david_2008_3.pdf)

**Comment by surajRathi** [comment]
![image](https://user-images.githubusercontent.com/6284428/114596529-61dbc480-9cad-11eb-8a6e-01b8b78a6232.png)

One circle is not good enough, need to implement with two circles.

Note: Using 
> Length: 192 in (4,877 mm)
> Width: 76 in (1,930 mm)
from [indylights.com](https://www.indylights.com/about/equipment-package)

**Comment by surajRathi** [comment]
dynamic_collision_checking was only checking the first two points in the given path, because it expected input in the format 3xn but  actually received a path_t (nx2)

**Comment by surajRathi** [comment]
New vehicle hit-box model: 
![image](https://user-images.githubusercontent.com/6284428/114669036-13ffa480-9d1f-11eb-8e53-ab67779010e3.png)

![image](https://user-images.githubusercontent.com/6284428/114669079-211c9380-9d1f-11eb-8dac-617e403303b4.png)

See: [desmos](https://www.desmos.com/calculator/meb2gdbtmg)

**Comment by surajRathi** [comment]
Now actually checks for collision at each point, 

![image](https://user-images.githubusercontent.com/6284428/114678962-5f1eb500-9d29-11eb-9ff5-69027bbb0cb3.png)

**Comment by surajRathi** [comment]
In c07d6fee139810139786f6a2e67800e55dcf0aca , now shift the other vehicle states to global frame only once, previously, states that were already shifted were being transformed again.

**Comment by surajRathi** [comment]
Issue should be resolved, needs testing in sim once.

**Comment by surajRathi** [comment]
Works in simulation

**Comment by surajRathi** [comment]
![image](https://user-images.githubusercontent.com/6284428/115177849-ab814080-a0ed-11eb-9b30-96c058453954.png)
No its not fixed, or maybe we need to actually generate better paths
